### PR TITLE
Use a card layout instead of a table for the dashboard on mobile

### PR DESCRIPTION
**Pull request by janfaracik**
**Before**
image

**After**
image

Updates the dashboard to use a list of cards rather than a table when on a mobile device (or device with a low enough resolution). The layout obviously isn't as dense (or information rich) but it should be a lot more usable when running on smaller devices.

### Testing done

* Open the dashboard on a large browser, notice the table view looks the same as it always has. Resize the browser to be small, notice the table has been replaced with a list of cards
* Open a [folder](https://plugins.jenkins.io/cloudbees-folder/), notice the same behaviour as above

### Proposed changelog entries

- Use a card layout instead of a table for the dashboard on mobile.

### Proposed upgrade guidelines

N/A

## Proposed developer guidelines
* Plugins can now include a compactColumn for the dashboard in mobile view

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7581"></a>

### CODE REVIEW

**Review by yaroslavafenkin — COMMENTED** [review]
Seems alright to me from security point of view.

File: war/src/main/less/base/visibility-utils.less
Comment (timja) [inline war/src/main/less/base/visibility-utils.less:170]: does this need documenting somewhere?

**Review by timja — COMMENTED** [review]
What do you think about including a Start build button (where applicable)?

Other than that looks really nice!

**Comment by daniel-beck** [comment]
Could support for this be retrofitted to the existing list view columns (e.g. they could have a new "compact representation")? Seems weird to have an entire branch that bypasses column configuration.

Alternatively it would be nicer to change or replace the existing "all view". A more opinionated alternative can do things like more tightly control what information gets shown without basically being tied to the list view/column model.

**Comment by janfaracik** [comment]
> Could support for this be retrofitted to the existing list view columns (e.g. they could have a new "compact representation")? Seems weird to have an entire branch that bypasses column configuration.

For the first iteration I've kept it simple, but I've definitely thought along the same lines as you on this - perhaps another row with compact statuses for the existing column headers.

> Alternatively it would be nicer to change or replace the existing "all view". A more opinionated alternative can do things like more tightly control what information gets shown without basically being tied to the list view/column model.

Definitely interested in this, what would you (and others) like to see different about the existing dashboard? I'd personally like it to be more dynamic rather than so fixed, but it's a trade off with customisability.

**Comment by uhafner** [comment]
> > Alternatively it would be nicer to change or replace the existing "all view". A more opinionated alternative can do things like more tightly control what information gets shown without basically being tied to the list view/column model.
> 
> Definitely interested in this, what would you (and others) like to see different about the existing dashboard? I'd personally like it to be more dynamic rather than so fixed, but it's a trade off with customisability.

I think we really need to replace the non-configurable all view with something that can be configured by a user (or team). The configuration options can be added incrementally. There are a lot of existing concepts that we can reuse: dahboard view, pull request monitoring view, jut to name a view. We can begin with the table and then head over to additional widgets that can be added and configured.

**Comment by daniel-beck** [comment]
> we really need to replace the non-configurable "all view" with something that can be configured by a user (or team). The configuration options can be added incrementally.

IMO it allowing no configuration isn't inherently bad, in fact it's the perfect location for having an opinionated default that easily adapts to e.g. screen resolution. It just being all items in no particular order with all the default columns is what makes it less useful than it could be IMO.

Folks with more complex needs who want something different can easily swap out the default view, or add additional ones.

File: war/src/main/less/base/visibility-utils.less
Comment (janfaracik) [reply war/src/main/less/base/visibility-utils.less:170]: Yeah, will add to the Design Library.

**Review by timja — COMMENTED** [review]
It looks really nice.

----

My only concern is that this applies to all views rather than just the default one, so if someone has lots of custom columns they won't be able to scroll and see their columns like they currently can, but unsure if it's a blocker

**Comment by janfaracik** [comment]
Updated it so that columns/plugins can provide a 'compactColumn.jelly' which will show when in card view.

image
_(in this case the column is returning the Jenkins logo and the value 1337)_

---

Needs to be documented etc.

File: war/src/main/less/base/visibility-utils.less
Comment (NotMyFault) [reply war/src/main/less/base/visibility-utils.less:170]: I don't mind a draft PR, for now.

**Review by NotMyFault — APPROVED** [review]
Nice work, should be documented within the design library, but looks remarkable!

---

> My only concern is that this applies to all views rather than just the default one, so if someone has lots of custom columns they won't be able to scroll and see their columns like they currently can

Preferably, you're using views and folders to group your jobs, no? Within these, you can scroll up and down on mobile still.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
